### PR TITLE
Use speakable type for top-level statements

### PR DESCRIPTION
**Pull request by jcouv**
I ran into an implementation problem: when SourceMemberContainerSymbol creates the non-type member symbols, it needs binders to hand them. But when we have top-level statements, some of the binder chains need to contain a binder for the entry point method (because top-level locals are in scope of the rest of the program, even though we'll disallow their usage). But we can't ask for the non-type member symbols to get that entry point method...
This PR breaks this cycle this PR computes the entry point symbols in an earlier phase, and stores them it a dedicated cache.

Fixes https://github.com/dotnet/roslyn/issues/54877

LDM:
- https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-07-12.md
- https://github.com/dotnet/csharplang/blob/main/meetings/2021/LDM-2021-07-26.md

### CODE REVIEW

File: src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs [hunk 1]
```diff
@@ -357,6 +357,6 @@ public static class WellKnownMemberNames
         /// <summary>
         /// The name of a type synthesized for a top-level statements entry point method.
         /// </summary>
-        public const string TopLevelStatementsEntryPointTypeName = "<Program>$";
+        public const string TopLevelStatementsEntryPointTypeName = "Program";
```
Comment (Youssef1313) [inline src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs:360]: FYI @davidwengier @tmat to see if this will break EnC:

https://github.com/dotnet/roslyn/blob/b46c7a752cfa94177dba13b836c7ccd5394d9043/src/Features/Core/Portable/EditAndContinue/AbstractEditAndContinueAnalyzer.cs#L5451-L5452

Comment (Youssef1313) [reply src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs:360]: Also FYI @CyrusNajmabadi, it's likely that generate field/property is going to break in a Program class:

https://github.com/dotnet/roslyn/blob/a07c7ff00fa58595250feb5c140193b9eba8d8d4/src/Features/Core/Portable/GenerateMember/GenerateVariable/AbstractGenerateVariableService.State.cs#L137-L140

Comment (davidwengier) [reply src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs:360]: Thanks @Youssef1313. EnC itself should be fine, but I would expect test failures which will be easy enough to correct

Comment (tmat) [reply src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs:360]: I don't think we can continue checking for top-level code container type just by name. We need to check if the `Program` type is in fact the synthesized implicit type (or if it is partial we need to somehow check if it has a synthesized partial declaration). It would be good to have a compiler API that tells us whether a type is the implicit top-level type.

There might also be some cases to test/consider around `partial class Program`.

Comment (jcouv) [reply src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs:360]: The distinctive mark of a simple program will be the `Program.<Main>$` method (remains unspeakable).

Comment (tmat) [reply src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs:360]: So we'll need to check if the type contains `<Main>$`? Seems like there should be API that does that.

Comment (jcouv) [reply src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs:360]: The code was already checking for `<Main>$` so it's not really a change.
We can discuss possible public API, but I'm not clear what would the user scenarios be. Do we just care whether there is ` <Main>$`? Do we care whether it's a pure simple program, or whether it has some user-defined members from a partial declaration?

Comment (tmat) [reply src/Compilers/Core/Portable/Symbols/WellKnownMemberNames.cs:360]: I see. I guess it's ok then. For some reason I thought we are checking the type.

File: src/EditorFeatures/CSharpTest/GenerateVariable/GenerateVariableTests.cs [hunk 2]
```diff
@@ -2511,6 +2511,17 @@ public static int ExMethod(this string s)
...
 
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateVariable)]
+        public async Task SpeakableTopLevelStatementType()
+        {
+            await TestMissingAsync(@"
+[|P|] = 10;
+
+partial class Program
+{
+}");
```
Comment (Youssef1313) [inline src/EditorFeatures/CSharpTest/GenerateVariable/GenerateVariableTests.cs:2522]: A more ideal behavior is to generate the property as follows:

```csharp
P = 10;

partial class Program
{
    public int P { get; private set; }
}
```

In case you find it out-of-scope for this PR, we can open a tracking issue for that.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 3]
```diff
@@ -2572,6 +2586,23 @@ internal void AddOrWrapTupleMembers(SourceMemberContainerTypeSymbol type)
             }
         }
 
+        protected sealed class SimpleProgramEntryPointInfo
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2589]: >protected [](http://example.com/codeflow?start=8&length=9)

`private`?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 4]
```diff
@@ -2911,8 +2942,74 @@ public void Free()
...
+            if (alreadyKnown != SimpleProgramEntryPointInfo.UninitializedSentinel)
+            {
+                diagnostics.Free();
+                return alreadyKnown?.SimpleProgramEntryPoints ?? ImmutableArray<SynthesizedSimpleProgramEntryPointSymbol>.Empty;
+            }
+
+            AddDeclarationDiagnostics(diagnostics);
+            diagnostics.Free();
+
+            return simpleProgramEntryPointInfo?.SimpleProgramEntryPoints ?? ImmutableArray<SynthesizedSimpleProgramEntryPointSymbol>.Empty;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2972]: Consider removing the duplicate `Free()` and `return`:
```csharp
if (alreadyKnown == SimpleProgramEntryPointInfo.UninitializedSentinel)
{
    AddDeclarationDiagonstics(diagnostics);
}
diagnostics.Free();
return _lazySimpleProgramEntryPoint?.SimpleProgramEntryPoints ??
    ImmutableArray<SynthesizedSimpleProgramEntryPointSymbol>.Empty;
```

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 5]
```diff
@@ -2911,8 +2942,74 @@ public void Free()
...
+            if (this.ContainingSymbol is not NamespaceSymbol { IsGlobalNamespace: true }
+                || this.Name != WellKnownMemberNames.TopLevelStatementsEntryPointTypeName)
+            {
+                return ImmutableArray<SynthesizedSimpleProgramEntryPointSymbol>.Empty;
+            }
+
+            SimpleProgramEntryPointInfo? simpleProgramEntryPointInfo = _lazySimpleProgramEntryPoint;
+            if (simpleProgramEntryPointInfo != SimpleProgramEntryPointInfo.UninitializedSentinel)
+            {
+                return simpleProgramEntryPointInfo?.SimpleProgramEntryPoints ?? ImmutableArray<SynthesizedSimpleProgramEntryPointSymbol>.Empty;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2956]: >return [](http://example.com/codeflow?start=16&length=6)

Consider changing the `if` to check `==` and moving the code below into the `if` block, so there is a single `return` since the return value is not trivial.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 6]
```diff
@@ -2911,8 +2942,74 @@ public void Free()
...
+            SimpleProgramEntryPointInfo? buildSimpleProgramEntryPoint(BindingDiagnosticBag diagnostics)
+            {
+                bool reportAnError = false;
+                ArrayBuilder<SynthesizedSimpleProgramEntryPointSymbol>? builder = null;
+
+                foreach (var singleDecl in declaration.Declarations)
+                {
+                    if (singleDecl.IsSimpleProgram)
+                    {
+                        if (reportAnError)
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2983]: >reportAnError [](http://example.com/codeflow?start=28&length=13)

Could `reportAnError` be replaced with `builder is { }`?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 7]
```diff
@@ -3455,6 +3552,15 @@ private void CheckForStructBadInitializers(DeclaredMembersAndInitializersBuilder
             }
         }
 
+        private void AddSynthesizedSimpleProgramEntryPointIfNecessary(MembersAndInitializersBuilder builder, DeclaredMembersAndInitializers declaredMembersAndInitializers, BindingDiagnosticBag diagnostics)
+        {
+            var simpleProgramEntryPoints = GetSimpleProgramEntryPoints();
+            foreach (var member in simpleProgramEntryPoints)
+            {
+                builder.AddNonTypeMember(member, declaredMembersAndInitializers);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3560]: >AddNonTypeMember(member, declaredMembersAndInitializers) [](http://example.com/codeflow?start=24&length=56)

Do we have a race condition where external callers of `GetSimpleProgramEntryPoints()` might get back an entry point that hasn't been added to `MembersAndInitializers` yet? If so, what are the implications?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs [hunk 8]
```diff
@@ -1572,5 +1578,14 @@ internal override bool Equals(TypeSymbol t2, TypeCompareKind comparison)
                 nativeInteger.Equals(this, comparison) :
                 base.Equals(t2, comparison);
         }
+
+#nullable enable
+        internal bool IsSimpleProgram
+        {
+            get
+            {
+                return this.declaration.Declarations.Any(d => d is { IsSimpleProgram: true });
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceNamedTypeSymbol.cs:1587]: >d is { IsSimpleProgram: true } [](http://example.com/codeflow?start=62&length=30)

Consider using `d.IsSimpleProgram` to match the other use above. Or replace both with a property on `MergedTypeDeclaration`.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs [hunk 9]
```diff
@@ -53,7 +55,7 @@ internal SynthesizedSimpleProgramEntryPointSymbol(SimpleProgramNamedTypeSymbol c
             bool isNullableAnalysisEnabled = IsNullableAnalysisEnabled(compilation, CompilationUnit);
             this.MakeFlags(
                 MethodKind.Ordinary,
-                DeclarationModifiers.Static | DeclarationModifiers.Private | (hasAwait ? DeclarationModifiers.Async : DeclarationModifiers.None),
+                DeclarationModifiers.Static | DeclarationModifiers.Public | (hasAwait ? DeclarationModifiers.Async : DeclarationModifiers.None),
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs:58]: >Public [](http://example.com/codeflow?start=67&length=6)

Does the main method need to be public?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs [hunk 10]
```diff
@@ -65,6 +67,37 @@ internal SynthesizedSimpleProgramEntryPointSymbol(SimpleProgramNamedTypeSymbol c
...
+        }
+
+        internal static SynthesizedSimpleProgramEntryPointSymbol? GetSimpleProgramEntryPoint(CSharpCompilation compilation)
+        {
+            return GetSimpleProgramNamedTypeSymbol(compilation)?.GetSimpleProgramEntryPoints().First();
+        }
+
+        private static SourceNamedTypeSymbol? GetSimpleProgramNamedTypeSymbol(CSharpCompilation compilation)
+        {
+            return compilation.SourceModule.GlobalNamespace.GetTypeMembers("Program").OfType<SourceNamedTypeSymbol>().SingleOrDefault(s => s.IsSimpleProgram);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs:98]: >"Program" [](http://example.com/codeflow?start=75&length=9)

`WellKnownMemberNames.TopLevelStatementsEntryPointName`

**Comment by cston** [comment]
>         public override Accessibility DeclaredAccessibility

Consider removing this override, rather than duplicating the value here, since `base.DeclaredAccessibility` uses the value set in the .ctor.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs:146 in 9dc9216. [](commit_id = 9dc9216df3e7af0734d26d64ce31d2c2428c2267, deletion_comment = False)

File: src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests_CallerArgumentExpression.vb [hunk 11]
```diff
@@ -134,7 +134,7 @@ End Class
             Dim source As String = "
 Imports System
 Imports System.Runtime.CompilerServices
-Public Module Program
+Public Module Program2
```
Comment (cston) [inline src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests_CallerArgumentExpression.vb:137]: >Program2 [](http://example.com/codeflow?start=14&length=8)

Is this change necessary?

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 7 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3560]: I don't think that is possible. `GetSimpleProgramEntryPoints` is atomic. It will return exactly one set of entry points, then we'll incorporate them into `MembersAndInitializers`. 
Different threads will see the same set of entry points, and there is not other way to get entry point symbols than through `GetSimpleProgramEntryPoints`.

Does that make sense, or do you still think there might be a way?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs (hunk 9 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedSimpleProgramEntryPointSymbol.cs:58]: We'd settled on public when we discussed a speakable method name. But we've reverted that decision since. I'll revert this to private, since it's unspeakable anyways.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 12]
```diff
@@ -8907,10 +8856,420 @@ class Program
...
-                // (7,17): warning CS7022: The entry point of the program is global code; ignoring 'Program.Main(string[])' entry point.
+                // (7,17): warning CS7022: The entry point of the program is global code; ignoring 'Program2.Main(string[])' entry point.
                 //     static void Main(String[] args) {}
-                Diagnostic(ErrorCode.WRN_MainIgnored, "Main").WithArguments("Program.Main(string[])").WithLocation(7, 17)
+                Diagnostic(ErrorCode.WRN_MainIgnored, "Main").WithArguments("Program2.Main(string[])").WithLocation(7, 17)
+                );
+        }
+
+        [Fact]
+        public void SpeakableEntyrPoint()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:8866]: >Entyr [](http://example.com/codeflow?start=29&length=5)

Typo.

File: src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests_CallerArgumentExpression.vb (hunk 11 above)
Comment (jcouv) [reply src/Compilers/VisualBasic/Test/Emit/Attributes/AttributeTests_CallerArgumentExpression.vb:137]: Yes. We're compiling a C# program below (on line 194) that needs to call the `Log` method from this VB program.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 13]
```diff
@@ -8907,10 +8856,420 @@ class Program
...
+
+            var comp = CreateCompilation(text, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+            CompileAndVerify(comp, expectedOutput: "Hi!");
+            var entryPoint = SynthesizedSimpleProgramEntryPointSymbol.GetSimpleProgramEntryPoint(comp);
+            Assert.Equal(Accessibility.Internal, entryPoint.ContainingType.DeclaredAccessibility);
+            Assert.Equal(Accessibility.Public, entryPoint.DeclaredAccessibility);
+        }
+
+        [Fact]
+        public void SpeakableEntyrPoint_ProgramIsStruct()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:9029]: >ProgramIsStruct [](http://example.com/codeflow?start=40&length=15)

Consider testing `partial interface Program`. And it looks like that test, along with the `struct` and `record` tests here could be combined with `[Theory]`.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 14]
```diff
@@ -8907,10 +8856,420 @@ class Program
...
+            var comp = CreateCompilation(text, options: TestOptions.DebugExe, parseOptions: DefaultParseOptions);
+            comp.VerifyDiagnostics(
+                // (8,9): error CS0103: The name 'Main' does not exist in the current context
+                //         Main();
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "Main").WithArguments("Main").WithLocation(8, 9)
+                );
+        }
+
+        [Fact]
+        public void SpeakableEntyrPoint_ProgramHasBaseList()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:9107]: >ProgramHasBaseList [](http://example.com/codeflow?start=40&length=18)

Consider testing with interfaces as well.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 15]
```diff
@@ -5130,84 +5131,6 @@ static void Main()
...
-    static void Main()
-    {
-        Console.Write(""hello "");
-    }
-
-    static async Task Main(string[] args)
-    {
-        await Task.Factory.StartNew(() => 5);
-        Console.Write(""async main"");
-    }
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:5193]: I'll restore this. It's just producing diagnostics for each `Main()` method.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 16]
```diff
@@ -8907,10 +8856,420 @@ class Program
...
+            var verifier = CompileAndVerify(comp, expectedOutput: "42");
+            verifier.VerifyDiagnostics();
+            var entryPoint = SynthesizedSimpleProgramEntryPointSymbol.GetSimpleProgramEntryPoint(comp);
+            Assert.Equal(Accessibility.Public, entryPoint.DeclaredAccessibility);
+            Assert.True(entryPoint.IsStatic);
+
+            Assert.Equal("Base", entryPoint.ContainingType.BaseType().ToTestDisplayString());
+            Assert.Equal(Accessibility.Internal, entryPoint.ContainingType.DeclaredAccessibility);
+            Assert.False(entryPoint.ContainingType.IsStatic);
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:9135]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing with:
- `abstract partial class Program { }`
- `sealed partial class Program { }`
- `static partial class Program { }`

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs (hunk 13 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:9029]: Interfaces would require a different test because of static method `M`... Not sure it's worth the trouble

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 17]
```diff
@@ -2911,8 +2942,69 @@ public void Free()
...
+                || this.Name != WellKnownMemberNames.TopLevelStatementsEntryPointTypeName)
+            {
+                return ImmutableArray<SynthesizedSimpleProgramEntryPointSymbol>.Empty;
+            }
+
+            SimpleProgramEntryPointInfo? simpleProgramEntryPointInfo = _lazySimpleProgramEntryPoint;
+            if (simpleProgramEntryPointInfo == SimpleProgramEntryPointInfo.UninitializedSentinel)
+            {
+
+                var diagnostics = BindingDiagnosticBag.GetInstance();
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2957]: >                 [](http://example.com/codeflow?start=0&length=16)

Consider removing blank line above.

**Review by cston — APPROVED** [review]

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 18]
```diff
@@ -182,6 +187,9 @@ public bool SetNullableContext(byte? value)
             TupleExtraData? tupleData = null)
             : base(tupleData)
         {
+            // If we're dealing with a simple program, then we must be in the global namespace
+            Debug.Assert(containingSymbol is NamespaceSymbol { IsGlobalNamespace: true } || !declaration.Declarations.Any(d => d.IsSimpleProgram));
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:191]: Is it true that we can't get here in an error scenario like the following:

```cs
namespace NS;

var x = 1;
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 19]
```diff
@@ -8907,10 +8936,553 @@ class Program
...
+            var cMember = comp.GetMember<NamedTypeSymbol>("Program");
+            Assert.Equal(
+@"<member name=""T:Program"">
+    <summary>Summary</summary>
+</member>
+", cMember.GetDocumentationCommentXml());
+        }
+
+        [Fact]
+        public void SpeakableEntryPoint_XmlDoc_ProgramIsPartial_NotCommented()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:9469]: nit: It feels like the motivating scenario was `typeof(Program).Assembly` and reflection. Consider testing this in a scenario where `partial class Program` is not explicitly declared.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 20]
```diff
@@ -2572,6 +2586,23 @@ internal void AddOrWrapTupleMembers(SourceMemberContainerTypeSymbol type)
             }
         }
 
+        private sealed class SimpleProgramEntryPointInfo
+        {
+            public readonly ImmutableArray<SynthesizedSimpleProgramEntryPointSymbol> SimpleProgramEntryPoints;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2591]: Is it true that in non-error scenarios, there are either 0 or 1 elements in here?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2591]: It feels like if we expect exactly one of these in a non-error scenario we might put a field on this class for that "single" one, and only use this field for error recovery. However, since it's not likely that very many of these will be created, it doesn't seem like an important change.

File: src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs [hunk 21]
```diff
@@ -1913,18 +1913,18 @@ private static bool CanBeInControlFlowGraph(IOperation n)
...
+            return symbol is IMethodSymbol
             {
                 Name: WellKnownMemberNames.TopLevelStatementsEntryPointMethodName,
                 ContainingType: { } containingType
-            } && containingType.IsTopLevelMainType();
-
-        private static bool IsTopLevelMainType([NotNullWhen(true)] this ISymbol? symbol)
-            => symbol is INamedTypeSymbol
+            }
+            && containingType is INamedTypeSymbol
```
Comment (RikkiGibson) [inline src/Compilers/Test/Core/Compilation/ControlFlowGraphVerifier.cs:1922]: nit: this seems like it could be a little more elegant as one big pattern :)

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 20 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2591]: Correct. There's one entry per file with top-level statement.
However, there's code (`SynthesizedSimpleProgramEntryPointSymbol? GetSimpleProgramEntryPoint(CSharpCompilation compilation, CompilationUnitSyntax compilationUnit, bool fallbackToMainEntryPoint)`) which needs to access all of the entry point symbols. That's why I couldn't store just zero/one here.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 18 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:191]: Correct. I'll add a few tests.
`ParseNamespaceBody` parses differently in global context (outside of a namespace). See `isGlobal` local. It recognizes when inside a namespace (either with brace or semi-colon).

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs (hunk 19 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:9469]: Good idea. Adding a typeof test.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 22]
```diff
@@ -2572,6 +2586,23 @@ internal void AddOrWrapTupleMembers(SourceMemberContainerTypeSymbol type)
             }
         }
 
+        private sealed class SimpleProgramEntryPointInfo
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2589]: >SimpleProgramEntryPointInfo [](http://example.com/codeflow?start=29&length=27)

Remove?

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs [hunk 23]
```diff
@@ -3393,7 +3393,7 @@ static void Test()
                 .True("Test();")
                 .True("Microsoft.CodeAnalysis.Runtime.Instrumentation.FlushPayload();")
                 .True(@"Console.WriteLine(""Test"");");
-            checker.Method(4, 1)
+            checker.Method(5, 1)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:3396]: >checker.Method(5, 1) [](http://example.com/codeflow?start=12&length=20)

Do we understand what method is added here?

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs [hunk 24]
```diff
@@ -4517,7 +4517,7 @@ public void TopLevelStatement_Closure()
             CheckEncLogDefinitions(reader1,
                 Row(2, TableIndex.StandAloneSig, EditAndContinueOperation.Default),
                 Row(1, TableIndex.MethodDef, EditAndContinueOperation.Default),
-                Row(3, TableIndex.MethodDef, EditAndContinueOperation.Default),
+                Row(4, TableIndex.MethodDef, EditAndContinueOperation.Default),
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/EditAndContinueClosureTests.cs:4520]: >4 [](http://example.com/codeflow?start=20&length=1)

Do we understand why this number has changed?

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 25]
```diff
@@ -5158,15 +5159,16 @@ static async Task Main()
...
+                Diagnostic(ErrorCode.ERR_MissingPartial, "Program").WithArguments("Program").WithLocation(7, 7),
                 // (9,17): warning CS7022: The entry point of the program is global code; ignoring 'Program.Main(string[])' entry point.
                 //     static void Main(string[] args)
                 Diagnostic(ErrorCode.WRN_MainIgnored, "Main").WithArguments("Program.Main(string[])").WithLocation(9, 17),
                 // (14,23): warning CS7022: The entry point of the program is global code; ignoring 'Program.Main()' entry point.
                 //     static async Task Main()
                 Diagnostic(ErrorCode.WRN_MainIgnored, "Main").WithArguments("Program.Main()").WithLocation(14, 23)
                 );
-
-            CompileAndVerify(comp, expectedOutput: "Hi!");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:5169]: >CompileAndVerify(comp, expectedOutput: "Hi!"); [](http://example.com/codeflow?start=12&length=46)

It looks like this test was testing interesting success scenario. I think we should keep it as such. The class name isn't important, so renaming it would be more appropriate.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 26]
```diff
@@ -5197,15 +5199,16 @@ static async Task Main(string[] args)
...
+                Diagnostic(ErrorCode.ERR_MissingPartial, "Program").WithArguments("Program").WithLocation(8, 7),
                 // (10,17): warning CS7022: The entry point of the program is global code; ignoring 'Program.Main()' entry point.
                 //     static void Main()
                 Diagnostic(ErrorCode.WRN_MainIgnored, "Main").WithArguments("Program.Main()").WithLocation(10, 17),
                 // (15,23): warning CS7022: The entry point of the program is global code; ignoring 'Program.Main(string[])' entry point.
                 //     static async Task Main(string[] args)
                 Diagnostic(ErrorCode.WRN_MainIgnored, "Main").WithArguments("Program.Main(string[])").WithLocation(15, 23)
                 );
-
-            CompileAndVerify(comp, expectedOutput: "Hi!");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:5208]: >CompileAndVerify(comp, expectedOutput: "Hi!"); [](http://example.com/codeflow?start=12&length=46)

Same here

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 27]
```diff
@@ -5263,20 +5266,13 @@ public static async Task Wait()
         public void ExplicitMain_10()
         {
             var text = @"
-using System.Threading.Tasks;
-
-System.Console.Write(""Hi!"");
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:5268]: Why the test scenario is drastically modified?

File: src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs (hunk 23 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Emit/Emit/DynamicAnalysis/DynamicInstrumentationTests.cs:3396]: We didn't use to generate a parameterless constructor for the entry-point type (which is a reference type). Now we do.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs (hunk 27 above)
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:5268]: Sure. We can rename the type name
